sunbird-rc-esignet-integration-impl/src/main/java/io/mosip/esignet/sunbird/integration/service/AuthenticationHelperService.java

```diff
@@ -0,0 +1,154 @@
+
+            for (AuthChallenge authChallenge : kycAuthDto.getChallengeList()) {
+                if(Objects.equals(authChallenge.getAuthFactorType(),"KBA")){
+                    return knowledgeBasedAuthentcation(kycAuthDto.getIndividualId(),authChallenge);
+                }
+                throw new KycAuthException("invalid_challenge_format");
+            }
+        } catch (KycAuthException e) {
+            throw e;
+        } catch (Exception e) {
+            log.error("KYC-auth failed with transactionId : {} && clientId : {}", kycAuthDto.getTransactionId(),
+                    clientId, e);
+        }
+        throw new KycAuthException(ErrorConstants.AUTH_FAILED);
+    }
+
+    private KycAuthResult knowledgeBasedAuthentcation(String policyNumber, AuthChallenge authChallenge) throws KycAuthException {
+
+        KycAuthResult  kycAuthResult= new KycAuthResult();
+        SearchRequestDto searchRequestDto=new SearchRequestDto();
+        searchRequestDto.setLimit(2);
+        searchRequestDto.setOffset(0);
+        Map<String,Map<String,String>> filter=new HashMap<>();
+        Map<String,String> fullNameMap=new HashMap<>();
+        Map<String,String> dobMap=new HashMap<>();
+        Map<String,String> policyMap=new HashMap<>();
+        String encodedChallenge=authChallenge.getChallenge();
+
+        byte[] base64Bytes = encodedChallenge.getBytes(StandardCharsets.UTF_8);
+        byte[] decodedBytes = Base64.getDecoder().decode(base64Bytes);
+        String challenge = new String(decodedBytes, StandardCharsets.UTF_8);
+
+        try {
+            ObjectMapper objectMapper = new ObjectMapper();
+            JsonNode jsonNode = objectMapper.readTree(challenge);
+
+            String fullName = jsonNode.get("name").asText();
+            String dob = jsonNode.get("dob").asText();
+
+
+            fullNameMap.put(fieldKey,fullName);
+            dobMap.put(fieldKey,dob);
+            policyMap.put(fieldKey,policyNumber);
+
+            filter.put(Full_Name,fullNameMap);
+            filter.put(DOB,dobMap);
+            filter.put(Policy_Number,policyMap);
+
+            searchRequestDto.setFilters(filter);
+
+            String requestBody = objectMapper.writeValueAsString(searchRequestDto);
+            RequestEntity requestEntity = RequestEntity
+                    .post(UriComponentsBuilder.fromUriString(kycAuthUrl).build().toUri())
+                    .contentType(MediaType.APPLICATION_JSON_UTF8)
+                    .body(requestBody);
+            ResponseEntity<List<InsurenceResponceDto>> responseEntity = restTemplate.exchange(requestEntity,
+                    new ParameterizedTypeReference<List<InsurenceResponceDto>>() {});
+            if (responseEntity.getStatusCode().is2xxSuccessful() && responseEntity.getBody() != null) {
+                List<InsurenceResponceDto> responseList = responseEntity.getBody();
+                if(responseList.size()==1){
```

Comment:
We should have a else block for this if condition to clearly log we got more than one response and what is the size. - as error log

Reply:
done